lib/CodeGen/CGExprScalar.cpp

```diff
@@ -2319,6 +2319,106 @@ static BinOpInfo createBinOpInfoFromIncDec(const UnaryOperator *E,
   return BinOp;
 }
 
+static Expr *peelOffPointerArithmetic(const BinaryOperator *B) {
+  if (B->isAdditiveOp() && B->getType()->isPointerType()) {
+    if (B->getLHS()->getType()->isPointerType()) {
+      return B->getLHS();
+    } else if (B->getRHS()->getType()->isPointerType()) {
+      return B->getRHS();
+    }
+  }
+  return nullptr;
+}
+
+static Expr *peelOffOuterExpr(Expr *E) {
```

Comment:
1. Can we rename this function to: skipCasts or something similar, which expresses the intent more clearly.
2. What about parenthesis? Do we create expressions for them in the ASTs as well? If so, we'd want to skip them as well.
3. The check for: if (const auto *BO = dyn_cast<BinaryOperator>(E)) can be more targeted, since you only handle additive operator. So, you can use: isAdditiveOp() here.

Reply:
1. `peelOffOuterExpr` has been adapted from a similar function in `lib/StaticAnalyzer/Core/BugReporterVisitors.cpp`. I wanted to keep the same name to make the association clear and also to help while grep'ping for the function.

2. Parenthesis are represented in the AST as node levels. For example: `(a * b) + c` is represented is:
```
                    +
                  /   \
                 *     c
               /   \
              a     b
```

3. `isAdditiveOp` is already checked inside `peelOffPointerArithmetic`.